File: IIFLapis/IIFLapis/order.py
"""
Contains base classes for Orders etc.
"""
from .const import GENERIC_PAYLOAD, ORDER_PAYLOAD, HEADERS, NEXT_DAY_TIMESTAMP
import requests
from .conf import APP_SOURCE
from enum import Enum


class Exchange(Enum):

    NSE = "N"
    BSE = "B"
    MCX = "M"


class ExchangeSegment(Enum):

    CASH = "C"
    DERIVATIVE = "D"
    CURRENCY = "U"


class OrderFor(Enum):

    PLACE = "P"
    MODIFY = "M"
    CANCEL = "C"


class OrderType(Enum):

    BUY = "BUY"
    SELL = "SELL"


class OrderValidity(Enum):

    DAY = 0
    GTD = 1
    GTC = 2
    IOC = 3
    EOS = 4
    FOK = 6


class AHPlaced(Enum):

    AFTER_MARKET_CLOSED = "Y"
    NORMAL_ORDER = "N"


class Order:

    def __init__(self, order_type: str, scrip_code: int, quantity: int, exchange: str,
                 exchange_segment: str, price: float ,is_intraday: bool , atmarket: bool ,order_id: int = 0,
                  remote_order_id: str = "1", exch_order_id: str = "0", order_for: OrderFor = OrderFor.PLACE,
                 DisQty: int=0, stoploss_price: float = 0, is_stoploss_order: bool = False, ioc_order: bool = False,
                  is_vtd: bool = False, vtd: str = f"/Date({NEXT_DAY_TIMESTAMP})/",
                 ahplaced: AHPlaced = AHPlaced.NORMAL_ORDER, public_ip: str = '192.168.1.1',
                 order_validity: OrderValidity = OrderValidity.DAY, traded_qty: int = 0):

        self.order_for = order_for.value
        self.exchange = exchange
        self.exchange_segment = exchange_segment
        self.price = price
        self.order_id = order_id
        self.order_type = order_type
        self.quantity = quantity
        self.scrip_code = scrip_code
        self.atmarket = atmarket
        self.remote_order_id = remote_order_id
        self.exch_order_id = exch_order_id
        self.DisQty=DisQty
        self.stoploss_price = stoploss_price
        self.is_stoploss_order = is_stoploss_order
        self.ioc_order = ioc_order
        self.is_intraday = is_intraday
        self.is_vtd = is_vtd
        self.vtd = vtd
        self.ahplaced = ahplaced.value
        self.public_ip = public_ip
        self.order_validity = order_validity.value
        self.traded_qty = traded_qty

File: Go lang/IIFL/class/Order.go
package class

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

//OrderRequest :- This API is the backbone of all the trade dependent APIs and is the key method that you would need to place your order in the market.
func OrderRequest(w http.ResponseWriter, r *http.Request) {
	ClientCode := r.URL.Query().Get("ClientCode")

	config := CheckConfig()
	url := config.ServiceURL + "OrderRequest"

	method := "POST"

	header := Header{AppName: config.AppName, AppVer: config.AppVer, Key: config.Key, OsName: config.OsName, RequestCode: config.RequestCodeOrdReq, UserID: config.UserID, Password: config.Password}

	body := OrderRequestBody{ClientCode: ClientCode, OrderFor: "P", Exchange: "N", ExchangeType: "C", Price: 190, OrderID: 1, OrderType: "BUY",
		Qty: 1, OrderDateTime: "/Date(1601880914379)/", ScripCode: 3045, AtMarket: false, RemoteOrderID: "s000220190", ExchOrderID: "0", DisQty: 0,
		IsStopLossOrder: false, StopLossPrice: 0, IsVTD: false, IOCOrder: false, IsIntraday: false, PublicIP: "192.168.84.215", AHPlaced: "N",
		ValidTillDate: "/Date(1602880914379)/", IOrderValidity: 0, OrderRequesterCode: "96130000", TradedQty: 0}

	OrderData := ReqDataRequest{Head: header, Body: body}

	OrderAPI := OrderRequestAPI{ReqData: OrderData, AppSource: 54}

	data, _ := json.Marshal(OrderAPI)

	payload := strings.NewReader(string(data))

	bodyString := HTTPClient(method, url, payload)
	json.NewEncoder(w).Encode(bodyString)

	var OrderRes OrderResponse

	// json to golang struct
	json.Unmarshal([]byte(bodyString), &OrderRes)
	fmt.Printf("Message: %s, StatusDescription: %s", OrderRes.Body.Message, OrderRes.Head.StatusDescription)

}

//OrderStatus :- The purpose of this method is to navigate the status of an order placed by the client.
func OrderStatus(w http.ResponseWriter, r *http.Request) {

	ClientCode := r.URL.Query().Get("ClientCode")

	config := CheckConfig()
	//fmt.Println(domainame.AppName)
	url := config.ServiceURL + "OrderStatus"

	method := "POST"

	header := Header{AppName: config.AppName, AppVer: config.AppVer, Key: config.Key, OsName: config.OsName, RequestCode: config.RequestCodeOrdStatus, UserID: config.UserID, Password: config.Password}

	OrdReqList := OrdStatusReqList{Exch: "N", ExchType: "C", ScripCode: 4717, RemoteOrderID: "S123456789123400000"}

	body := OrderStatusBody{ClientCode: ClientCode, OrdStatusReqList: []OrdStatusReqList{OrdReqList}}

	OrderBookV2Request := OrderStatusRequest{Head: header, Body: body}

	data, _ := json.Marshal(OrderBookV2Request)

	payload := strings.NewReader(string(data))

	bodyString := HTTPClient(method, url, payload)
	json.NewEncoder(w).Encode(bodyString)

	var OrderStatusRes OrderStatusResponse

	// json to golang struct
	json.Unmarshal([]byte(bodyString), &OrderStatusRes)
	fmt.Printf("Message: %s, StatusDescription: %s", OrderStatusRes.Body.Message, OrderStatusRes.Head.StatusDescription)

}

//TradeInformation :- Trade Information method is used to Fetch Trade Information for a set of orders placed by the client.
func TradeInformation(w http.ResponseWriter, r *http.Request) {

	ClientCode := r.URL.Query().Get("ClientCode")

	config := CheckConfig()

	url := config.ServiceURL + "TradeInformation"

	method := "POST"

	header := Header{AppName: config.AppName, AppVer: config.AppVer, Key: config.Key, OsName: config.OsName, RequestCode: config.RequestCodeTrdInfo, UserID: config.UserID, Password: config.Password}

	TradeInformList := OrdStatusReqList{Exch: "B", ExchType: "C", ScripCode: 500410, RemoteOrderID: "1557728588259000000"}

	body := OrderStatusBody{ClientCode: ClientCode, OrdStatusReqList: []OrdStatusReqList{TradeInformList}}

	OrderBookV2Request := OrderStatusRequest{Head: header, Body: body}

	data, _ := json.Marshal(OrderBookV2Request)

	payload := strings.NewReader(string(data))

	bodyString := HTTPClient(method, url, payload)
	json.NewEncoder(w).Encode(bodyString)

	var TradeInformationRes TradeInformationResponse

	// json to golang struct
	json.Unmarshal([]byte(bodyString), &TradeInformationRes)
	fmt.Printf("Message: %s, StatusDescription: %s", TradeInformationRes.Body.Message, TradeInformationRes.Head.StatusDescription)

}
